public/app/features/dashboard/dashgrid/PanelChrome.tsx

```diff
@@ -67,6 +79,51 @@ export class PanelChrome extends PureComponent<Props, State> {
     this.props.panel.events.off('refresh', this.onRefresh);
   }
 
+  // Updates the response with information from the stream
+  panelDataObserver = {
+    next: (event: PanelDataEvent) => {
+      const data = {
+        ...this.state.data,
+        ...event,
+      };
+
+      if (data.state === LoadingState.Error) {
```

Comment:
Would be cool if such things could be handled via some kind of hooks rather than implemented inline in the observer. `PanelDataEvent` could expose an API for that, i.e. 
`event.whenSuccessful(...).whenInProgress(...)` or sth like that.

Reply:
totally agree -- the structure of this is a little funny because the intended use is not actually in this panel :) it is written like this so it is easy to understand the change we are making, but the Subject & Observable are a little silly if they just sit in the same component talking to each other!